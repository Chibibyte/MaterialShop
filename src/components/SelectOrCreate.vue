<template>
  <div>
    <v-row class="justify-space-around">
      <v-col cols="12">
        <v-btn v-if="updateContent" @click="() => openAddData()">
          <span>Create new</span>
          <v-icon>mdi-plus</v-icon>
        </v-btn>
        <v-radio-group v-model="activeRadio">
          <v-hover v-for="data in dataArr" :key="dataToKeyFunction(data)">
            <template v-slot="{ hover }">
              <v-card hover :elevation="hover ? 24 : 6" class="mb-10">
                <v-row>
                  <v-col cols="9">
                    <slot name="main" :data="data">SLOTFALLBACK</slot>
                    <v-container
                      class="d-flex justify-space-around"
                      v-if="updateContent"
                    >
                      <v-btn @click="() => openUpdateData(data)">
                        <v-icon>mdi-pencil-outline</v-icon>
                      </v-btn>
                      <v-btn @click="() => openRemoveData(data)">
                        <v-icon>mdi-delete-outline</v-icon>
                      </v-btn>
                    </v-container>
                  </v-col>
                  <v-divider vertical inset></v-divider>

                  <v-col
                    @click="() => selectRadio(data)"
                    class="d-flex justify-center align-items-center pa-0"
                  >
                    <v-radio
                      label="Select"
                      :ref="() => getRadioRef(data)"
                      :value="getContentRef(data)"
                      @change="selectData(data)"
                    ></v-radio>
                  </v-col>
                </v-row>
              </v-card>
            </template>
          </v-hover>
        </v-radio-group>
      </v-col>
    </v-row>

    <!-- DIALOG -->
    <template v-if="updateContent">
      <SelectOrCreateDialog
        v-for="(dialog, name) in dialogs"
        :key="name"
        v-model="dialogs[name]"
        :title="dataDialogText[name + 'Text']"
        :confirmFunction="
          () => {
            if (name === 'add') addData(dataDialogData);
            else if (name === 'remove') removeData(dataDialogData);
            else if (name === 'update')
              updateData(dataDialogOldData, dataDialogData);
          }
        "
      >
        <template #default="slotProps">
          <slot
            :validate="slotProps.validate"
            :name="name"
            :data="dataDialogData"
            :callback="copyObj"
            :disabled="name === 'remove' ? true : false"
            >{{ name.toUpperCase() }}DIALOG FALLBACK</slot
          >
        </template>
      </SelectOrCreateDialog>
    </template>
  </div>
</template>

<script>
/**
 * Business Logic
 *    -component for multiple different CRUD-Operations on state data
 *
 * Inputs:
 *    -updateContent: determines wheather CRUD-Operations can be performed on this content by showing/hiding elements
 *    -contentArrName:  name of the data source array in the vuex store
 *    -dataToKeyFunction: function that returns a key generates with the current items data
 *    -chState: name of the state in the vuex store for the current data set
 *    -emptyDataFunction: function that returns an object of the current data type with keys but empty values (e.g. {var1: "", var2: -1},
 *                        depends on what you want an empty object of your data to look like instead of just undefined)
 *    -dataDialogText: Object containing the different texts for the different crud dialogs (e.g. 'Delete Account', 'Add Address')
 * Outputs:
 *    -dataDialogText
 *    -crudData (contentMethodType, content)
 */
function propMissing(propName) {
  let text = propName + "MISSING";
  // console.error(text);
  return text;
}
import SelectOrCreateDialog from "@/components/SelectOrCreateDataDialog.vue";
export default {
  name: "SelectOrCreate",
  computed: {
    console: () => console,
  },
  components: {
    SelectOrCreateDialog,
  },
  props: {
    verboseMissingProps: {
      type: Boolean,
      default: false,
    },
    updateContent: {
      type: Boolean,
      default: true,
    },
    contentArrName: {
      type: String,
      default: "NO DATA ARRAY",
    },
    dataToKeyFunction: {
      type: Function,
      default: (data) =>
        Object.values(data).reduce((acc, cur) => acc + cur.toString()),
    },
    chState: {
      type: String,
      default: () => propMissing("chState"),
    },

    emptyDataFunction: {
      type: Function,
      default: () => propMissing("emptyDataFunction"),
    },
    dataDialogText: {
      type: Object,
      default() {
        return {
          addText: "add",
          removeText: "remove",
          updateText: "update",
        };
      },
    },
  },
  data() {
    return {
      dataArr: this.$store.state.checkout[this.contentArrName],
      checkoutStates: this.$store.state.checkout.checkoutStates,
      activeRadio: "",
      // data dialog
      dataDialogData: this.emptyDataFunction(),

      // toggle dialog
      dialogs: {
        add: false,
        remove: false,
        update: false,
      },

      dataDialogOldData: this.emptyDataFunction(),

      dialogTypes: ["add", "update", "delete"],
    };
  },
  watch: {
    dataDialogData: {
      handler() {
        this.$emit("input", this.dataDialogData);
      },
      deep: true,
    },
  },
  methods: {
    getContentRef(data) {
      return this.contentArrName + "-ref-" + this.dataToKeyFunction(data);
    },
    getRadioRef(data) {
      return this.contentArrName + "-radioref-" + this.dataToKeyFunction(data);
    },
    copyObj(objSource, altTarget) {
      Object.keys(objSource).forEach((key) => {
        if (altTarget) altTarget[key] = objSource[key];
        else this.dataDialogData[key] = objSource[key];
      });
    },
    selectRadio(data) {
      let radioValue =
        this.contentArrName + "-ref-" + this.dataToKeyFunction(data);
      this.activeRadio = radioValue;
      this.selectData(data);
    },
    selectData(data) {
      this.$store.commit("checkout/selectContent", {
        contentName: this.chState,
        content: data,
      });
      this.$store.commit("checkout/setCheckoutState", {
        chState: this.chState,
        value: true,
      });
    },

    openUpdateData(data) {
      this.copyObj(data);
      this.copyObj(data, this.dataDialogOldData);
      this.dialogs["update"] = true;
    },
    openAddData() {
      this.copyObj(this.emptyDataFunction());
      this.dialogs["add"] = true;
    },

    openRemoveData(data) {
      this.copyObj(data);
      this.copyObj(data, this.dataDialogOldData);
      this.dialogs["remove"] = true;
    },
    crudData(contentMethodType, content) {
      if (this.contentArrName == "NO DATA ARRAY") return;
      this.$store.commit(`checkout/${contentMethodType}Content`, {
        contentArrName: this.contentArrName,
        content,
      });
      this[contentMethodType + "Dialog"] = false;
    },
    addData(content) {
      this.crudData("add", content);
    },
    removeData(content) {
      this.crudData("remove", content);
    },
    updateData(oldData, data) {
      let content = [oldData, data];
      this.crudData("update", content);
    },
  },
  mounted() {
    if (this.$store.state.checkout[this.chState])
      this.selectRadio(this.$store.state.checkout[this.chState]);
  },
};
</script>