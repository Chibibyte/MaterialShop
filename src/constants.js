export default {
    BUTTONICONS: {
        AUTHENTICATION: 'mdi-login-variant',
        ADDRESS: 'mdi-package-variant',
        BANKACCOUNT: 'mdi-bank',
        SHIPPING: 'mdi-truck',
        PAYMENT: 'mdi-cash',
        SUMMARY: 'mdi-text-box-check-outline',
        CHECKOUT_COMPLETE: 'mdi-check',
        ADD: 'mdi-plus',
        REMOVE: 'mdi-minus',
        CANCEL: 'mdi-window-close',
        CONFIRM: 'mdi-check',
        SHOPPING: 'mdi-shopping',
        SUCCESS: 'mdi-check',
        FAILURE: 'mdi-window-close',
        WARNING: 'mdi-alert-circle-outline'
    }
}