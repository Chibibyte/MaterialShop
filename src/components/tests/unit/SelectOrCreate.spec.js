import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils';
import SelectOrCreate from '@/components/SelectOrCreate.vue';
import constants from '@/constants.js'

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.prototype.$constants = constants;

let stubs = [
    'v-container',
    'v-row',
    'v-col',
    'v-radio-group',
    'v-btn',
    'v-icon',
    'v-hover'
]

let wrapper;
let state;
let modules;
let store;
function createBankaccount(bank, accountNr) {
    return {
        bank,
        accountNr
    };
}
beforeEach(() => {
    state = {
    };
    modules = {
        checkout: {
            namespaced: true,
            state() {
                return {
                    bankaccounts: [
                        createBankaccount('Bank-A', '1234'),
                        createBankaccount('Bank-B', '5678'),
                        createBankaccount('Bank-C', '1357'),
                        createBankaccount('Bank-D', '2468')
                    ],
                    checkoutStates: {
                        authentication: false,
                        deliveryAddress: false,
                        bankaccount: false,
                        shipping: false,
                        payment: false,
                        summary: false,
                    },
                }
            },
            mutations: {
                addContent: jest.fn(),
                removeContent: jest.fn(),
                updateContent: jest.fn()
            }
        }
    }
    store = new Vuex.Store({
        state,
        modules
    })
    wrapper = shallowMount(SelectOrCreate, { stubs, localVue, store });
})

describe('Component', () => {
    test('is instantiated', () => {
        expect(wrapper.isVueInstance).toBeTruthy();
    })

    test('default snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    })
})

describe('props:updateContent', () => {
    test('default:true:no data', () => {
        expect(wrapper.vm.updateContent).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
    })

    test('false:no data', () => {
        wrapper.setProps({ updateContent: false });
        expect(wrapper.vm.updateContent).toBe(false);
        expect(wrapper.element).toMatchSnapshot();
    })

    test('default:true:bankaccounts data', () => {
        wrapper = shallowMount(SelectOrCreate, { stubs, localVue, store, propsData: { contentArrName: 'bankaccounts' } });
        expect(wrapper.vm.updateContent).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
    })

    test('false:bankaccounts data', () => {
        wrapper.setProps({ contentArrName: 'bankaccounts', updateContent: false });
        expect(wrapper.vm.updateContent).toBe(false);
        expect(wrapper.element).toMatchSnapshot();
    })
})

describe('props:dataDialogText', () => {
    test('default:', () => {
        expect(wrapper.vm.dataDialogText.addText).toBe('add');
        expect(wrapper.vm.dataDialogText.removeText).toBe('remove');
        expect(wrapper.vm.dataDialogText.updateText).toBe('update');
    })

    test('testTexts', () => {
        wrapper.setProps({ dataDialogText: { addText: 'test add', removeText: 'test delete', updateText: 'test update' } })
        expect(wrapper.vm.dataDialogText.addText).toBe('test add');
        expect(wrapper.vm.dataDialogText.removeText).toBe('test delete');
        expect(wrapper.vm.dataDialogText.updateText).toBe('test update');
    })
})

describe('data:dataDialogData', () => {
    test('default:', () => {
        expect(wrapper.vm.dataDialogData).toBe("emptyDataFunctionMISSING");
    })

    test('{a:"testA", b:"testB"}:', async () => {
        wrapper.vm.dataDialogData = { a: "testA", b: "testB" };
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.dataDialogData).toEqual({ a: "testA", b: "testB" });
        expect(wrapper.emitted().input).toBeTruthy();
        expect(wrapper.emitted().input[0][0]).toEqual({ a: "testA", b: "testB" });
    })
})

describe('methods:crudData', () => {
    test('default: addContent', () => {
        wrapper.vm.crudData('add', { varA: 'valueA', varB: 'valueB' });
        expect(modules.checkout.mutations.addContent.mock.calls).toHaveLength(0);
    })

    function contentArrTest(crudType) {
        let methodType = crudType + 'Content';
        test(`contentArrName === "bankaccounts": ${methodType}`, () => {
            wrapper = shallowMount(SelectOrCreate, { stubs, localVue, store, propsData: { contentArrName: 'bankaccounts' } });
            wrapper.vm.crudData(crudType, { varA: 'valueA', varB: 'valueB' });
            expect(modules.checkout.mutations[methodType].mock.calls).toHaveLength(1);
            expect(modules.checkout.mutations[methodType].mock.calls[0][1]).toEqual({ content: { varA: 'valueA', varB: 'valueB' }, contentArrName: 'bankaccounts' });
        })
    }

    contentArrTest('add');
    contentArrTest('remove');
    contentArrTest('update');

})