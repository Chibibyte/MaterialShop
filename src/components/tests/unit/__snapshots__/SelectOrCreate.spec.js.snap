// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component default snapshot 1`] = `
<div>
  <v-row-stub
    class="justify-space-around"
  >
    <v-col-stub
      cols="12"
    >
      <v-btn-stub>
        <span>
          Create new
        </span>
         
        <v-icon-stub>
          add
        </v-icon-stub>
      </v-btn-stub>
       
      <v-radio-group-stub
        value=""
      />
    </v-col-stub>
  </v-row-stub>
   
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="add"
  />
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="remove"
  />
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="update"
  />
</div>
`;

exports[`props:updateContent default:true:bankaccounts data 1`] = `
<div>
  <v-row-stub
    class="justify-space-around"
  >
    <v-col-stub
      cols="12"
    >
      <v-btn-stub>
        <span>
          Create new
        </span>
         
        <v-icon-stub>
          add
        </v-icon-stub>
      </v-btn-stub>
       
      <v-radio-group-stub
        value=""
      >
        <v-hover-stub />
        <v-hover-stub />
        <v-hover-stub />
        <v-hover-stub />
      </v-radio-group-stub>
    </v-col-stub>
  </v-row-stub>
   
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="add"
  />
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="remove"
  />
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="update"
  />
</div>
`;

exports[`props:updateContent default:true:no data 1`] = `
<div>
  <v-row-stub
    class="justify-space-around"
  >
    <v-col-stub
      cols="12"
    >
      <v-btn-stub>
        <span>
          Create new
        </span>
         
        <v-icon-stub>
          add
        </v-icon-stub>
      </v-btn-stub>
       
      <v-radio-group-stub
        value=""
      />
    </v-col-stub>
  </v-row-stub>
   
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="add"
  />
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="remove"
  />
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="update"
  />
</div>
`;

exports[`props:updateContent false:bankaccounts data 1`] = `
<div>
  <v-row-stub
    class="justify-space-around"
  >
    <v-col-stub
      cols="12"
    >
      <v-btn-stub>
        <span>
          Create new
        </span>
         
        <v-icon-stub>
          add
        </v-icon-stub>
      </v-btn-stub>
       
      <v-radio-group-stub
        value=""
      />
    </v-col-stub>
  </v-row-stub>
   
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="add"
  />
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="remove"
  />
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="update"
  />
</div>
`;

exports[`props:updateContent false:no data 1`] = `
<div>
  <v-row-stub
    class="justify-space-around"
  >
    <v-col-stub
      cols="12"
    >
      <v-btn-stub>
        <span>
          Create new
        </span>
         
        <v-icon-stub>
          add
        </v-icon-stub>
      </v-btn-stub>
       
      <v-radio-group-stub
        value=""
      />
    </v-col-stub>
  </v-row-stub>
   
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="add"
  />
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="remove"
  />
  <selectorcreatedialog-stub
    confirmfunction="function () {
        if(name === 'add') { _vm.addData(_vm.dataDialogData) }
        else if(name === 'remove') { _vm.removeData(_vm.dataDialogData) }
        else if(name === 'update') { _vm.updateData(_vm.dataDialogOldData, _vm.dataDialogData) }
        }"
    title="update"
  />
</div>
`;
